common/src/main/java/io/netty/util/NetUtil.java

```diff
@@ -934,6 +935,25 @@ public static Inet6Address getByName(CharSequence ip, boolean ipv4Mapped) {
     }
 
     /**
+     * Returns the {@link String} representation of an {@link InetSocketAddress}.
+     * <p>
+     * The output does not include Scope ID.
+     * @param addr {@link InetSocketAddress} to be converted to an address string
+     * @return {@code String} containing the text-formatted IP address
+     */
+    public static String toSocketAddressString(InetSocketAddress addr) {
+        final String rhost;
+        if (addr.isUnresolved()) {
+            String hostString = PlatformDependent.javaVersion() >= 7 ? addr.getHostString() : addr.getHostName();
+            rhost = isValidIpV6Address(hostString) ? '[' + hostString + ']' : hostString;
+        } else {
+            InetAddress address = addr.getAddress();
+            rhost = address instanceof Inet4Address ? toAddressString(address) : '[' + toAddressString(address) + ']';
+        }
+        return rhost + ':' + addr.getPort();
```

Comment:
I wonder if it would should be more explicit about the string concatenation operations used here. For example it seems unlikely JIT will be able to combine all of these conditional concatenation operations into a single `StringBuilder`. Can we create a single `StringBuilder` and manually do the `append` calls (e.g. including what is now building `rhost`) to be sure we don't create multiple `String`/`StringBuilder` operations?